sklearn/tree/_criterion.pyx

```diff
@@ -962,6 +969,304 @@ cdef class MSE(RegressionCriterion):
         impurity_left[0] /= self.n_outputs
         impurity_right[0] /= self.n_outputs
 
+cdef class MAE(RegressionCriterion):
+    """Mean absolute error impurity criterion"""
+    def __dealloc__(self):
+        """Destructor."""
+
+        free(self.node_medians)
+
+    cdef np.ndarray left_child_heaps
+    cdef np.ndarray right_child_heaps
+    cdef double* node_medians
+
+    def __cinit__(self, SIZE_t n_outputs):
+        """Initialize parameters for this criterion.
+
+        Parameters
+        ----------
+        n_outputs: SIZE_t
+            The number of targets to be predicted
+        """
+
+        # Default values
+        self.y = NULL
+        self.y_stride = 0
+        self.sample_weight = NULL
+
+        self.samples = NULL
+        self.start = 0
+        self.pos = 0
+        self.end = 0
+
+        self.n_outputs = n_outputs
+        self.n_node_samples = 0
+        self.weighted_n_node_samples = 0.0
+        self.weighted_n_left = 0.0
+        self.weighted_n_right = 0.0
+
+        # Allocate accumulators. Make sure they are NULL, not uninitialized,
+        # before an exception can be raised (which triggers __dealloc__).
+        self.node_medians = NULL
+
+        # Allocate memory for the accumulators
+        self.node_medians = <double*> calloc(n_outputs, sizeof(double))
+
+        if (self.node_medians == NULL):
+            raise MemoryError()
+
+        self.left_child_heaps = np.empty(n_outputs, dtype='object')
+        self.right_child_heaps = np.empty(n_outputs, dtype='object')
```

Comment:
Hmm I actually haven't tried that, cython doesn't like arrays of python
objects so I was using this hack Jacob showed me. I'll try out using
pointers directly and let you know if there are any issues with it.

On Thu, Jun 23, 2016, 3:11 PM Gilles Louppe <notifications@github.com>
wrote:

> In sklearn/tree/_criterion.pyx
> <https://github.com/scikit-learn/scikit-learn/pull/6667#discussion_r68185222>
> :
>
> > +        self.weighted_n_node_samples = 0.0
> > +        self.weighted_n_left = 0.0
> > +        self.weighted_n_right = 0.0
> > +
> > +        # Allocate accumulators. Make sure they are NULL, not uninitialized,
> > +        # before an exception can be raised (which triggers __dealloc__).
> > +        self.node_medians = NULL
> > +
> > +        # Allocate memory for the accumulators
> > +        self.node_medians = <double*> calloc(n_outputs, sizeof(double))
> > +
> > +        if (self.node_medians == NULL):
> > +            raise MemoryError()
> > +
> > +        self.left_child_heaps = np.empty(n_outputs, dtype='object')
> > +        self.right_child_heaps = np.empty(n_outputs, dtype='object')
>
> Why a numpy array of objects, instead of a MedianHeap*? This would avoid
> a lot of casting all throughout the implementation.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6667/files/20d6107c7ebfba1b96d483cfacb97a9ac304b3cb#r68185222>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AG72XzHaDyfg0OfWF-wIL05rZ9qeryZgks5qOjGVgaJpZM4IIwnV>
> .
>